File: examples/go/merge_intervals.go
package main

import (
	"fmt"
	"sort"
)

/*
Merge Intervals Pattern Examples

The Merge Intervals pattern is used to deal with overlapping intervals.
It involves sorting intervals and then merging or manipulating them based on overlap.

Time Complexity: O(n log n) due to sorting
Space Complexity: O(n) for output
*/

// MergeIntervals merges overlapping intervals
// Time: O(n log n), Space: O(n)
func MergeIntervals(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}

	// Sort intervals by start time
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	merged := [][]int{intervals[0]}

	for i := 1; i < len(intervals); i++ {
		current := intervals[i]
		lastMerged := merged[len(merged)-1]

		// If current interval overlaps with last merged interval
		if current[0] <= lastMerged[1] {
			// Merge intervals by updating end time
			if current[1] > lastMerged[1] {
				lastMerged[1] = current[1]
			}
		} else {
			// No overlap, add current interval
			merged = append(merged, current)
		}
	}

	return merged
}

// InsertInterval inserts a new interval into a sorted list of non-overlapping intervals
// Time: O(n), Space: O(n)
func InsertInterval(intervals [][]int, newInterval []int) [][]int {
	result := [][]int{}
	i := 0

	// Add all intervals that end before new interval starts
	for i < len(intervals) && intervals[i][1] < newInterval[0] {
		result = append(result, intervals[i])
		i++
	}

	// Merge overlapping intervals with new interval
	for i < len(intervals) && intervals[i][0] <= newInterval[1] {
		if intervals[i][0] < newInterval[0] {
			newInterval[0] = intervals[i][0]
		}
		if intervals[i][1] > newInterval[1] {
			newInterval[1] = intervals[i][1]
		}
		i++
	}

	result = append(result, newInterval)

	// Add remaining intervals
	for i < len(intervals) {
		result = append(result, intervals[i])
		i++
	}

	return result
}

// IntervalIntersection finds intersection of two lists of intervals
// Time: O(m + n), Space: O(min(m, n))
func IntervalIntersection(firstList, secondList [][]int) [][]int {
	result := [][]int{}
	i, j := 0, 0

	for i < len(firstList) && j < len(secondList) {
		// Find intersection
		start := max(firstList[i][0], secondList[j][0])
		end := min(firstList[i][1], secondList[j][1])

		// If valid intersection exists
		if start <= end {
			result = append(result, []int{start, end})
		}

		// Move pointer of interval that ends first
		if firstList[i][1] < secondList[j][1] {
			i++
		} else {
			j++
		}
	}

	return result
}

// CanAttendMeetings checks if a person can attend all meetings
// Time: O(n log n), Space: O(1)
func CanAttendMeetings(intervals [][]int) bool {
	if len(intervals) == 0 {
		return true
	}

	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] < intervals[i-1][1] {
			return false
		}
	}

	return true
}

// MinMeetingRooms finds minimum number of meeting rooms required
// Time: O(n log n), Space: O(n)
func MinMeetingRooms(intervals [][]int) int {
	if len(intervals) == 0 {
		return 0
	}

	// Create separate slices for start and end times
	starts := make([]int, len(intervals))
	ends := make([]int, len(intervals))

	for i, interval := range intervals {
		starts[i] = interval[0]
		ends[i] = interval[1]
	}

	sort.Ints(starts)
	sort.Ints(ends)

	roomsNeeded := 0
	maxRooms := 0
	startPtr, endPtr := 0, 0

	for startPtr < len(intervals) {
		// If a meeting starts before another ends, we need a new room
		if starts[startPtr] < ends[endPtr] {
			roomsNeeded++
			if roomsNeeded > maxRooms {
				maxRooms = roomsNeeded
			}
			startPtr++
		} else {
			// A meeting ended, free up a room
			roomsNeeded--
			endPtr++
		}
	}

	return maxRooms
}

// EmployeeFreeTime finds common free time for all employees
// Time: O(n log n), Space: O(n)
func EmployeeFreeTime(schedule [][][]int) [][]int {
	// Flatten all intervals
	allIntervals := [][]int{}
	for _, employeeSchedule := range schedule {
		allIntervals = append(allIntervals, employeeSchedule...)
	}

	// Merge all intervals
	merged := MergeIntervals(allIntervals)

	// Find gaps between merged intervals
	freeTime := [][]int{}
	for i := 1; i < len(merged); i++ {
		if merged[i-1][1] < merged[i][0] {
			freeTime = append(freeTime, []int{merged[i-1][1], merged[i][0]})
		}
	}

	return freeTime
}

// Helper functions
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// Helper function to check if two 2D slices are equal
func equal2D(a, b [][]int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if len(a[i]) != len(b[i]) {
			return false
		}
		for j := range a[i] {
			if a[i][j] != b[i][j] {
				return false
			}
		}
	}
	return true
}

// Test functions
func testMergeIntervals() {
	fmt.Println("Testing merge intervals...")

	testCases := []struct {
		input    [][]int
		expected [][]int
	}{
		{[][]int{{1, 3}, {2, 6}, {8, 10}, {15, 18}}, [][]int{{1, 6}, {8, 10}, {15, 18}}},
		{[][]int{{1, 4}, {4, 5}}, [][]int{{1, 5}}},
		{[][]int{{1, 4}, {0, 4}}, [][]int{{0, 4}}},
		{[][]int{{1, 4}, {2, 3}}, [][]int{{1, 4}}},
	}

	for _, tc := range testCases {
		result := MergeIntervals(tc.input)
		if !equal2D(result, tc.expected) {
			panic(fmt.Sprintf("Expected %v, got %v", tc.expected, result))
		}
	}

	fmt.Println("✅ Merge intervals tests passed")
}

func testInsertInterval() {
	fmt.Println("Testing insert interval...")

	testCases := []struct {
		intervals   [][]int
		newInterval []int
		expected    [][]int
	}{
		{[][]int{{1, 3}, {6, 9}}, []int{2, 5}, [][]int{{1, 5}, {6, 9}}},
		{[][]int{{1, 2}, {3, 5}, {6, 7}, {8, 10}, {12, 16}}, []int{4, 8}, [][]int{{1, 2}, {3, 10}, {12, 16}}},
		{[][]int{}, []int{5, 7}, [][]int{{5, 7}}},
		{[][]int{{1, 5}}, []int{2, 3}, [][]int{{1, 5}}},
	}

	for _, tc := range testCases {
		result := InsertInterval(tc.intervals, tc.newInterval)
		if !equal2D(result, tc.expected) {
			panic(fmt.Sprintf("Expected %v, got %v", tc.expected, result))
		}
	}

	fmt.Println("✅ Insert interval tests passed")
}

func testIntervalIntersection() {
	fmt.Println("Testing interval intersection...")

	first := [][]int{{0, 2}, {5, 10}, {13, 23}, {24, 25}}
	second := [][]int{{1, 5}, {8, 12}, {15, 24}, {25, 26}}
	expected := [][]int{{1, 2}, {5, 5}, {8, 10}, {15, 23}, {24, 24}, {25, 25}}

	result := IntervalIntersection(first, second)
	if !equal2D(result, expected) {
		panic(fmt.Sprintf("Expected %v, got %v", expected, result))
	}

	fmt.Println("✅ Interval intersection tests passed")
}

func testMeetingRooms() {
	fmt.Println("Testing meeting room problems...")

	// Can attend all meetings
	if CanAttendMeetings([][]int{{0, 30}, {5, 10}, {15, 20}}) != false {
		panic("Expected false for overlapping meetings")
	}
	if CanAttendMeetings([][]int{{7, 10}, {2, 4}}) != true {
		panic("Expected true for non-overlapping meetings")
	}

	// Minimum meeting rooms
	if MinMeetingRooms([][]int{{0, 30}, {5, 10}, {15, 20}}) != 2 {
		panic("Expected 2 meeting rooms")
	}
	if MinMeetingRooms([][]int{{7, 10}, {2, 4}}) != 1 {
		panic("Expected 1 meeting room")
	}
	if MinMeetingRooms([][]int{{9, 10}, {4, 9}, {4, 17}}) != 2 {
		panic("Expected 2 meeting rooms")
	}

	fmt.Println("✅ Meeting room tests passed")
}

func testEmployeeFreeTime() {
	fmt.Println("Testing employee free time...")

	schedule := [][][]int{
		{{1, 3}, {6, 7}},
		{{2, 4}},
		{{2, 5}, {9, 12}},
	}
	expected := [][]int{{5, 6}, {7, 9}}

	result := EmployeeFreeTime(schedule)
	if !equal2D(result, expected) {
		panic(fmt.Sprintf("Expected %v, got %v", expected, result))
	}

	fmt.Println("✅ Employee free time tests passed")
}

func main() {
	fmt.Println("Testing Merge Intervals Pattern...")
	testMergeIntervals()
	testInsertInterval()
	testIntervalIntersection()
	testMeetingRooms()
	testEmployeeFreeTime()
	fmt.Println("\n🎉 All Merge Intervals tests passed!")
}

File: examples/python/merge_intervals.py
#!/usr/bin/env python3
"""
Merge Intervals Pattern Examples

The Merge Intervals pattern is used to deal with overlapping intervals.
It involves sorting intervals and then merging or manipulating them based on overlap.

Time Complexity: O(n log n) due to sorting
Space Complexity: O(n) for output
"""

from typing import List


def merge_intervals(intervals: List[List[int]]) -> List[List[int]]:
    """
    Merge overlapping intervals.
    
    Args:
        intervals: List of intervals [start, end]
        
    Returns:
        List of merged intervals
        
    Time: O(n log n), Space: O(n)
    """
    if not intervals:
        return []
    
    # Sort intervals by start time
    intervals.sort(key=lambda x: x[0])
    merged = [intervals[0]]
    
    for current in intervals[1:]:
        last_merged = merged[-1]
        
        # If current interval overlaps with last merged interval
        if current[0] <= last_merged[1]:
            # Merge intervals by updating end time
            last_merged[1] = max(last_merged[1], current[1])
        else:
            # No overlap, add current interval
            merged.append(current)
    
    return merged


def insert_interval(intervals: List[List[int]], new_interval: List[int]) -> List[List[int]]:
    """
    Insert a new interval into a sorted list of non-overlapping intervals.
    
    Args:
        intervals: Sorted list of non-overlapping intervals
        new_interval: New interval to insert
        
    Returns:
        List of intervals after insertion and merging
        
    Time: O(n), Space: O(n)
    """
    result = []
    i = 0
    
    # Add all intervals that end before new interval starts
    while i < len(intervals) and intervals[i][1] < new_interval[0]:
        result.append(intervals[i])
        i += 1
    
    # Merge overlapping intervals with new interval
    while i < len(intervals) and intervals[i][0] <= new_interval[1]:
        new_interval[0] = min(new_interval[0], intervals[i][0])
        new_interval[1] = max(new_interval[1], intervals[i][1])
        i += 1
    
    result.append(new_interval)
    
    # Add remaining intervals
    while i < len(intervals):
        result.append(intervals[i])
        i += 1
    
    return result


def interval_intersection(first_list: List[List[int]], second_list: List[List[int]]) -> List[List[int]]:
    """
    Find intersection of two lists of intervals.
    
    Args:
        first_list: First list of intervals
        second_list: Second list of intervals
        
    Returns:
        List of intersection intervals
        
    Time: O(m + n), Space: O(min(m, n))
    """
    result = []
    i = j = 0
    
    while i < len(first_list) and j < len(second_list):
        # Find intersection
        start = max(first_list[i][0], second_list[j][0])
        end = min(first_list[i][1], second_list[j][1])
        
        # If valid intersection exists
        if start <= end:
            result.append([start, end])
        
        # Move pointer of interval that ends first
        if first_list[i][1] < second_list[j][1]:
            i += 1
        else:
            j += 1
    
    return result


def can_attend_meetings(intervals: List[List[int]]) -> bool:
    """
    Check if a person can attend all meetings (no overlapping intervals).
    
    Args:
        intervals: List of meeting intervals [start, end]
        
    Returns:
        True if can attend all meetings, False otherwise
        
    Time: O(n log n), Space: O(1)
    """
    if not intervals:
        return True
    
    intervals.sort(key=lambda x: x[0])
    
    for i in range(1, len(intervals)):
        if intervals[i][0] < intervals[i-1][1]:
            return False
    
    return True


def min_meeting_rooms(intervals: List[List[int]]) -> int:
    """
    Find minimum number of meeting rooms required.
    
    Args:
        intervals: List of meeting intervals [start, end]
        
    Returns:
        Minimum number of rooms needed
        
    Time: O(n log n), Space: O(n)
    """
    if not intervals:
        return 0
    
    # Create separate lists for start and end times
    starts = sorted([interval[0] for interval in intervals])
    ends = sorted([interval[1] for interval in intervals])
    
    rooms_needed = 0
    max_rooms = 0
    start_ptr = end_ptr = 0
    
    while start_ptr < len(intervals):
        # If a meeting starts before another ends, we need a new room
        if starts[start_ptr] < ends[end_ptr]:
            rooms_needed += 1
            max_rooms = max(max_rooms, rooms_needed)
            start_ptr += 1
        else:
            # A meeting ended, free up a room
            rooms_needed -= 1
            end_ptr += 1
    
    return max_rooms


def merge_intervals_with_names(intervals: List[List]) -> List[List]:
    """
    Merge intervals that also contain names/identifiers.
    
    Args:
        intervals: List of [start, end, name] intervals
        
    Returns:
        List of merged intervals with combined names
        
    Time: O(n log n), Space: O(n)
    """
    if not intervals:
        return []
    
    # Sort by start time
    intervals.sort(key=lambda x: x[0])
    merged = [intervals[0].copy()]
    
    for current in intervals[1:]:
        last_merged = merged[-1]
        
        # If current interval overlaps with last merged interval
        if current[0] <= last_merged[1]:
            # Merge intervals
            last_merged[1] = max(last_merged[1], current[1])
            # Combine names
            if len(current) > 2 and len(last_merged) > 2:
                last_merged[2] = f"{last_merged[2]},{current[2]}"
        else:
            # No overlap, add current interval
            merged.append(current.copy())
    
    return merged


def employee_free_time(schedule: List[List[List[int]]]) -> List[List[int]]:
    """
    Find common free time for all employees.
    
    Args:
        schedule: List of employee schedules, each containing intervals
        
    Returns:
        List of free time intervals common to all employees
        
    Time: O(n log n), Space: O(n)
    """
    # Flatten all intervals
    all_intervals = []
    for employee_schedule in schedule:
        all_intervals.extend(employee_schedule)
    
    # Merge all intervals
    merged = merge_intervals(all_intervals)
    
    # Find gaps between merged intervals
    free_time = []
    for i in range(1, len(merged)):
        if merged[i-1][1] < merged[i][0]:
            free_time.append([merged[i-1][1], merged[i][0]])
    
    return free_time


# Test functions
def test_merge_intervals():
    """Test merging overlapping intervals"""
    test_cases = [
        ([[1,3],[2,6],[8,10],[15,18]], [[1,6],[8,10],[15,18]]),
        ([[1,4],[4,5]], [[1,5]]),
        ([[1,4],[0,4]], [[0,4]]),
        ([[1,4],[2,3]], [[1,4]])
    ]
    
    for intervals, expected in test_cases:
        result = merge_intervals(intervals)
        assert result == expected, f"Expected {expected}, got {result}"
    
    print("✅ Merge intervals tests passed")


def test_insert_interval():
    """Test inserting interval"""
    test_cases = [
        ([[1,3],[6,9]], [2,5], [[1,5],[6,9]]),
        ([[1,2],[3,5],[6,7],[8,10],[12,16]], [4,8], [[1,2],[3,10],[12,16]]),
        ([], [5,7], [[5,7]]),
        ([[1,5]], [2,3], [[1,5]])
    ]
    
    for intervals, new_interval, expected in test_cases:
        result = insert_interval(intervals, new_interval)
        assert result == expected, f"Expected {expected}, got {result}"
    
    print("✅ Insert interval tests passed")


def test_interval_intersection():
    """Test interval intersection"""
    first = [[0,2],[5,10],[13,23],[24,25]]
    second = [[1,5],[8,12],[15,24],[25,26]]
    expected = [[1,2],[5,5],[8,10],[15,23],[24,24],[25,25]]
    
    result = interval_intersection(first, second)
    assert result == expected, f"Expected {expected}, got {result}"
    
    print("✅ Interval intersection tests passed")


def test_meeting_rooms():
    """Test meeting room problems"""
    # Can attend all meetings
    assert can_attend_meetings([[0,30],[5,10],[15,20]]) == False
    assert can_attend_meetings([[7,10],[2,4]]) == True
    
    # Minimum meeting rooms
    assert min_meeting_rooms([[0,30],[5,10],[15,20]]) == 2
    assert min_meeting_rooms([[7,10],[2,4]]) == 1
    assert min_meeting_rooms([[9,10],[4,9],[4,17]]) == 2
    
    print("✅ Meeting room tests passed")


def test_employee_free_time():
    """Test employee free time"""
    schedule = [[[1,3],[6,7]],[[2,4]],[[2,5],[9,12]]]
    expected = [[5,6],[7,9]]
    
    result = employee_free_time(schedule)
    assert result == expected, f"Expected {expected}, got {result}"
    
    print("✅ Employee free time tests passed")


if __name__ == "__main__":
    print("Testing Merge Intervals Pattern...")
    test_merge_intervals()
    test_insert_interval()
    test_interval_intersection()
    test_meeting_rooms()
    test_employee_free_time()
    print("\n🎉 All Merge Intervals tests passed!")